renku/core/metadata/repository.py

```diff
@@ -0,0 +1,1192 @@
+        super().__init__()
+        self._repository: git.Repo = repository or _create_repository(path, search_parent_directories)
+        self._path = Path(self._repository.working_dir).resolve()
+
+    @classmethod
+    def clone_from(
+        cls,
+        url: Union[Path, str],
+        path: Union[Path, str],
+        *,
+        branch: str = None,
+        recursive: bool = False,
+        depth: int = None,
+        progress: Optional[Callable] = None,
+        no_checkout: bool = False,
+        env: dict = None,
+    ) -> "Repository":
+        """Clone a remote repository and create an instance."""
+        try:
+            repository = git.Repo.clone_from(
+                url=url,
+                to_path=path,
+                branch=branch,
+                recursive=recursive,
+                depth=depth,
+                progress=progress,
+                no_checkout=no_checkout,
+                env=env,
+            )
+        except git.GitCommandError as e:
+            raise errors.GitCommandError(
+                message=f"Git command failed: {str(e)}",
+                command=e.command,
+                stdout=e.stdout,
+                stderr=e.stderr,
+                status=e.status,
+            ) from e
+        else:
+            return cls(path=path, repository=repository)
+
+    @classmethod
+    def initialize(cls, path: Union[Path, str], *, bare: bool = False, branch: str = None) -> "Repository":
+        """Initialize a git repository."""
+        try:
+            Path(path).mkdir(parents=True, exist_ok=True)
+            repository = git.Repo.init(path=str(path), bare=bare, initial_branch=branch)
+        except git.GitCommandError as e:
+            raise errors.GitCommandError(
+                message=f"Git command failed: {str(e)}",
+                command=e.command,
+                stdout=e.stdout,
+                stderr=e.stderr,
+                status=e.status,
+            ) from e
+        except git.GitError as e:
+            raise errors.GitError(f"Git command failed: {str(e)}") from e
+        else:
+            return cls(path=path, repository=repository)
+
+
+class Submodule(BaseRepository):
+    """A git submodule."""
+
+    def __init__(self, parent: git.Repo, name: str, path: Union[Path, str], url: str):
+        super().__init__(path=path, repository=None)
+        self._parent: git.Repo = parent or _create_repository(path=path, search_parent_directories=True)
+        self._name: str = name
+        self._url: str = url
+        try:
+            self._repository: git.Repo = _create_repository(path, search_parent_directories=False)
+        except errors.GitError:
+            # NOTE: Submodule directory doesn't exists yet, so, we ignore the error
+            pass
+
+    @classmethod
+    def from_submodule(cls, parent: git.Repo, submodule: Optional[git.Submodule]) -> Optional["Submodule"]:
+        """Create an instance from a git submodule."""
+        if submodule is None:
```

Comment:
Is submodule ever actually None? I didn't see any obvious place where it would be None